proposed/2021/VulnerabilityInfoInPMUI.md

```diff
@@ -0,0 +1,52 @@
+* Maintain the functionality of the deprecated warning icon in the package list as well as the installed tab. The icon will be used to indicate vulnerabilities, but will need to still indicate deprecated packages.
+* There can be one or more vulnerabilities for a package version so the details pane will need to be able to display a list of vulnerabilities, the severity levels, and the advisory links for each.
+
+## Goals
+
+* Show a warning icon in the Installed tab with a tooltip indicates how many vulnerable packages are installed.
+* Show a warning icon next to the package version of a vulnerable package in the package list. This icon will show for packages in the Browse and Installed tabs.
+* Show detailed information about the package vulnerabilities in the package details pane when the user clicks on a package.
+* Track via telemetry when an action (installed, update, uninstall) is performed on a package with a vulnerability from within the Package Manager UI.
+
+## Non-Goals
+
+* Improve the performance of querying package vulnerability information.
+* Change the architecture of the PMUI to not require querying information for the Installed tab to know if there is a deprecated package or vulnerability for the warning icon, and once again for the packages list. The existing architecture of the PMUI will be used and existing patterns will be followed.
+* Make any improvements to the package deprecation experience.
+* Change the registration API to pull in more details about a package vulnerability. This will be considered in a future iteration.
+* Show a warning icon in the Solution Explorer when packages with vulnerabilities are installed. This will be considered in a future iteration.
+
+## Solution
+
+A list of vulnerabilities for a package is read from the server, along with other data, whenever a package version is selected (DetailControlModel.SetCurrentPackageAsync) or when the selection version changes (DetailControlModel.SelectedVersionChangedAsync). This data is stored in the DetailedPackageMetadata and is used by the Details Pane. The vulnerabilities information exposed in the PackageSolutionDetailControlModel which implements DetailedPackageMetadata is currently limited to whether a vulnerability exists and what the maximum severity of the vulnerability is. This will be changed to expose instead a list of vulnerabilities, their severity, and their advisory links.
+
+The Details Pane will have a new section in the view called "Vulnerabilities" that will be visible when a package has one or more vulnerabilities. The Vulnerabilities section will display the highest severity of the vulnerabilities and show a list of vulnerabilities, their severity, and a link to the advisory. To stay consistent with the current design of the details pane, the vulnerabilities section will be created as its own UserControl that will be used by the DetailControl.
```

Comment:
Since it would be relatively rare that this section be shown, a perf consideration would be to make sure it doesn't try to render/bind or fire propertyChanged notifications for these new controls unnecessarily. 

We do have some perf issues where package `SelectionChanged` causes so many bindings in the details pane, we overwhelm some screen readers (and likely contributes negatively to perf in general). Something to be mindful of.

Reply:
I would imagine any issues with the bindings are more likely a problem with the property being bound to than the binding itself. For example, a binding should not be made to a Lazy property because it forces the evaluation of the Lazy to happen on the UI thread and I know we have a few cases where this is being done. If such issues are present, it would likely cause performance issues. I will be avoiding any such pitfalls. If you mean the bindings themselves of collapsed controls are causing performance issues, can you show me a perf trace?

Collapsed controls will not be picked up by screen readers. Can you expand on what you mean that bindings overwhelming screen readers?

Reply:
> Collapsed controls will not be picked up by screen readers. Can you expand on what you mean that bindings overwhelm screen readers?

I don't see where the spec indicates they will be collapsed. Either way, a collapsed control like a TreeView, for example, still fires UIA events when the binding changes (eg, our Dependencies TreeView). This is to notify any clients of the new data in that control.  

Simply put: please just make sure controls aren't being unnecessarily made visible and rebinded. Preferably, make them non-visible (or whatever the control specifies) to prevent unnecessary rebinding.

> Can you expand on what you mean that bindings overwhelm screen readers?

[NVDA becomes sluggish with WPF TreeView UIAutomation Events #11109](https://github.com/nvaccess/nvda/issues/11109)

Reply:
Rest assured I will not be making the vulnerabilities section visible when there are no vulnerabilities to display as that would be quite undesirable. :-)

>I don't see where the spec indicates they will be collapsed.

The spec says the control will be visible when there are vulnerabilities, implying it will not be visible otherwise. In WPF when a control is not visible and you want it to not take up space, you make it collapsed. I feel adding that level of detail would be too much about how WPF works.

Reply:
> The spec says the control will be visible when there are vulnerabilities,

Ah I see the confusion. Visible isn't the same thing as collapsed. If you mean `Visibility.Collapsed`, you could put that in the spec, because others like myself may think you're talking about a collapsed state of a control, such as a TreeView or the like.